libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+    // @User: Advanced
+    AP_GROUPINFO("W_ACTION",   2, AP_Avoidance, _warn_action, MAV_COLLISION_ACTION_REPORT),
+
+    // @Param: RCVRY_W
+    // @DisplayName: Recovery behaviour after a warn event
+    // @Description: Determines what the aircraft will do after a warn event (which has not progressed to a fail event) has been resolved
+    // @Values: 0:Resume previous flight mode,1:Mode Loiter,2:Continue collision avoidance action
+    // @User: Advanced
+    AP_GROUPINFO("W_RCVRY",   3, AP_Avoidance, _warn_recovery, AVOIDANCE_RECOVERY_W_RESUME),
+
+    // @Param: RCVRY_F
+    // @DisplayName: Recovery behaviour after a fail event
+    // @Description: Determines what the aircraft will do after a fail event is resolved
+    // @Values: 0:Continue collision avoidance ACTION_F,1:Move to collision avoidance W_ACTION
+    // @User: Advanced
+    AP_GROUPINFO("F_RCVRY",   4, AP_Avoidance, _fail_recovery, AVOIDANCE_RECOVERY_F_MOVE_TO_WARN),
+
+    // @Param: OBS_MAX
+    // @DisplayName: Maximum number of obstacles to track
+    // @Description: Maximum number of obstacles to track
+    // @User: Advanced
+    AP_GROUPINFO("OBS_MAX",   5, AP_Avoidance, _samples_max, 20),
+
+    // @Param: TIME_W_XY
+    // @DisplayName: Time Horizon Warn XY
+    // @Description: Aircraft horizontal velocity vectors are multiplied by this time to determine closest approach.  If this results in an approach closer than DIST_W_XY then BEHAVIOUR_W is undertaken (assuming BEHAVIOUR_F is not undertaken)
+    // @User: Advanced
+    AP_GROUPINFO("W_TIME_XY",     6, AP_Avoidance, _warn_time_horizon_xy,    30),
+
+    // @Param: TIME_F_XY
+    // @DisplayName: Time Horizon Warn XY
+    // @Description: Aircraft horizontal velocity vectors are multiplied by this time to determine closest approach.  If this results in an approach closer than DIST_F_XY then BEHAVIOUR_F is undertaken
+    // @User: Advanced
+    AP_GROUPINFO("F_TIME_XY",     7, AP_Avoidance, _fail_time_horizon_xy,    30),
+
+    // @Param: DIST_W_XY
+    // @DisplayName: Distance Warn XY
+    // @Description: Closest allowed projected distance before BEHAVIOUR_W is undertaken
+    // @User: Advanced
+    AP_GROUPINFO("W_DIST_XY",     8, AP_Avoidance, _warn_distance_xy,    300),
+
+    // @Param: DIST_F_XY
+    // @DisplayName: Distance Fail XY
+    // @Description: Closest allowed projected distance before BEHAVIOUR_F is undertaken
+    // @User: Advanced
+    AP_GROUPINFO("F_DIST_XY",     9, AP_Avoidance, _fail_distance_xy,    100),
+
+    AP_GROUPEND
+};
+
+const char *AP_Avoidance::_state_names[] = { "CLEAR", "WARN", "FAIL" };
+
+AP_Avoidance::AP_Avoidance(AP_AHRS &ahrs, AP_ADSB &adsb) :
+    _ahrs(ahrs),
+    _adsb(adsb)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _instance = this;
+}
+
+/*
+ * Initialize variables and allocate memory for array
+ */
+void AP_Avoidance::init(void)
+{
+    if (_samples == NULL) {
+        _samples = new AP_Avoidance::Sample[_samples_max];
```

Comment:
I suggest you make a copy of _samples_max on init and use that as the max in case AVOID_OBS_MAX changes. Then you can also detect if _samples_max changes and run deinit() and re-init() on the fly